lib/ddtrace/context.rb

```diff
@@ -111,12 +120,18 @@ def close_span(span)
         # on per-instrumentation code to retrieve handle parent/child relations.
         set_current_span(span.parent)
         return if span.tracer.nil?
-        return unless Datadog::Tracer.debug_logging
         if span.parent.nil? && !check_finished_spans
-          opened_spans = @trace.length - @finished_spans
-          Datadog::Tracer.log.debug("root span #{span.name} closed but has #{opened_spans} unfinished spans:")
-          @trace.each do |s|
-            Datadog::Tracer.log.debug("unfinished span: #{s}") unless s.finished?
+          if Datadog::Tracer.debug_logging
+            opened_spans = @trace.length - @finished_spans
+            Datadog::Tracer.log.debug("root span #{span.name} closed but has #{opened_spans} unfinished spans:")
+          end
+
+          @trace.reject(&:finished?).group_by(&:name).each do |unfinished_span_name, unfinished_spans|
```

Comment:
there is a chance there are a ton here, if you can think of like:

```python
# sorry for doing Python, quicker/easier for me to demonstrate

# Start root span
root = tracer.trace('root')

# Start a bunch of children
for i in range(1000):
    child = tracer.trace(f'child_{i}')

# Close root span, while all the child are still open
root.finish()
```

Would this then emit 1000 metrics?

There is a chance this could be spammy, maybe just once per-trace is enough to know the number of traces that are finishing with open children?

Maybe `Diagnostics::Health.metrics.error_unfinished_spans(unfinished_spans.length, tags: ["name:#{root_span.name}"])`

Where `unfinished_spans` is all of the unfinished spans, not grouped by their name?

Reply:
In your example, yeah, it'd emit 1000, because the `name` there is variable. That's atypical, incorrect usage, so while its possible, its not likely. 

Typically in this case if you had 1000 unfinished spans named `child`, that would not emit 1000 metrics, just 1, as it groups all the spans by `name`.

Tagging this with the root span name would be worthless... especially because almost all Ruby applications always have the same root span name (e.g. `rack.request`.) I think the possible extreme edge case is worth the risk in this case, because its the difference between having to randomly guess what span was bad in an entire trace versus knowing exactly which one had a problem.

Reply:
Yeah, just wondering what the overhead of doing that group by will be if there are a ton of unfinished spans and if it is worth it.

Reply:
If there's a ton of unfinished spans, then its definitely worth knowing about I'd say.

Reply:
I think this metric is definitely useful, we should 100% put something here, just wondering about like:

```ruby
# Emit a histogram for the number of unfinished spans per trace
unfinished_spans = @trace.reject(&:finished?)
statsd.histogram('unfinished_spans', unfinished_spans.length)
```

vs

```ruby
# Emit a metric for every unfinished span by name
@trace.reject(&:finished?).group_by(&:name).each do |unfinished_span_name, unfinished_spans|
    statsd.increment('unfinished_spans', unfinished_spans.length, tags: ["name:#{unfinished_span_name}"])
end
```

vs 

```ruby
# Just say there is a trace with unfinished spans, don't try to count them
statsd.increment('traces.with.unfinished_spans')
```

Just curious which approach here gives us enough information without being too costly to compute. I am concerned about a scenario where, lets say every single trace has like 100 unfinished spans.

Reply:
Just knowing that there are unfinished spans without know what spans are unfinished is insufficient to resolve the issue. The only time this is costly is when there's a ton of uniquely named spans that are all unfinished, and happens repeatedly.

This is very unlikely, but in the case that it does happen, then we'll know there's something terribly wrong in that application and how to fix it. If we only have the number of unfinished spans, we might know something is terribly wrong, but we'll have no idea how to fix it. The name of the unfinished span is critical here.

Reply:
I see unfinished spans as a critical condition in the application, and thus I think that having all the information we need to make this debug metric actionable is only reasonable, even at a performance penalty when this condition happens.